File: app/src/main/java/com/veyndan/paper/reddit/api/reddit/network/RedditService.java
package com.veyndan.paper.reddit.api.reddit.network;

import com.veyndan.paper.reddit.api.reddit.model.Listing;
import com.veyndan.paper.reddit.api.reddit.model.Thing;

import java.util.List;
import java.util.Map;

import io.reactivex.Single;
import retrofit2.Response;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.QueryMap;

public interface RedditService {

    // ================================
    //         Links & Comments
    // ================================

    @FormUrlEncoded
    @POST("api/hide")
    Single<Response<Void>> hide(
            @Field("id") String ids);

    @FormUrlEncoded
    @POST("api/save")
    Single<Response<Void>> save(
            @Field("category") String category,
            @Field("id") String id);

    @FormUrlEncoded
    @POST("api/unsave")
    Single<Response<Void>> unsave(
            @Field("id") String id);

    @FormUrlEncoded
    @POST("api/vote")
    Single<Response<Void>> vote(
            @Field("dir") VoteDirection voteDirection,
            @Field("id") String id);

    // ================================
    //             Listings
    // ================================

    @GET("r/{subreddit}/comments/{article}")
    Single<Response<List<Thing<Listing>>>> subredditComments(
            @Path("subreddit") String subreddit,
            @Path("article") String article);

    @GET("r/{subreddit}/{where}")
    Single<Response<Thing<Listing>>> subreddit(
            @Path("subreddit") String subreddit,
            @Path("where") Sort sort,
            @QueryMap Map<String, String> queries);

    // ================================
    //              Search
    // ================================

    @GET("r/{subreddit}/search")
    Single<Response<Thing<Listing>>> search(
            @Path("subreddit") String subreddit,
            @QueryMap Map<String, String> queries);

    // ================================
    //              Users
    // ================================

    @GET("user/{username}/{where}")
    Single<Response<Thing<Listing>>> user(
            @Path("username") String username,
            @Path("where") User where,
            @QueryMap Map<String, String> queries);
}

File: app/src/main/java/com/veyndan/paper/reddit/api/reddit/network/RedditService.kt
package com.veyndan.paper.reddit.api.reddit.network

import com.veyndan.paper.reddit.api.reddit.model.Listing
import com.veyndan.paper.reddit.api.reddit.model.Thing
import io.reactivex.Single
import retrofit2.Response
import retrofit2.http.*

interface RedditService {

    // ================================
    //         Links & Comments
    // ================================

    @FormUrlEncoded
    @POST("api/hide")
    fun hide(
            @Field("id") ids: String): Single<Response<Void>>

    @FormUrlEncoded
    @POST("api/save")
    fun save(
            @Field("category") category: String,
            @Field("id") id: String): Single<Response<Void>>

    @FormUrlEncoded
    @POST("api/unsave")
    fun unsave(
            @Field("id") id: String): Single<Response<Void>>

    @FormUrlEncoded
    @POST("api/vote")
    fun vote(
            @Field("dir") voteDirection: VoteDirection,
            @Field("id") id: String): Single<Response<Void>>

    // ================================
    //             Listings
    // ================================

    @GET("r/{subreddit}/comments/{article}")
    fun subredditComments(
            @Path("subreddit") subreddit: String,
            @Path("article") article: String): Single<Response<List<Thing<Listing>>>>

    @GET("r/{subreddit}/{where}")
    fun subreddit(
            @Path("subreddit") subreddit: String,
            @Path("where") sort: Sort,
            @QueryMap queries: Map<String, String>): Single<Response<Thing<Listing>>>

    // ================================
    //              Search
    // ================================

    @GET("r/{subreddit}/search")
    fun search(
            @Path("subreddit") subreddit: String,
            @QueryMap queries: Map<String, String>): Single<Response<Thing<Listing>>>

    // ================================
    //              Users
    // ================================

    @GET("user/{username}/{where}")
    fun user(
            @Path("username") username: String,
            @Path("where") where: User,
            @QueryMap queries: Map<String, String>): Single<Response<Thing<Listing>>>
}
